File: sdks/python/tyk_fapi/http_client.py
"""
HTTP client implementation for the Tyk FAPI SDK.
"""

import requests
from requests.adapters import HTTPAdapter
from requests.models import PreparedRequest
from typing import Dict, Optional, Any, Union
from urllib.parse import urlparse

from .token_source import TokenSource


class DPoPTransport(HTTPAdapter):
    """
    A transport adapter that adds DPoP proofs to requests.
    """
    
    def __init__(self, client, token_source: TokenSource):
        """
        Initialize the DPoP transport.
        
        Args:
            client: The Client instance to use for DPoP proof generation.
            token_source: The TokenSource to use for token acquisition.
        """
        super().__init__()
        self.client = client
        self.token_source = token_source
    
    def add_headers(self, request: PreparedRequest, **kwargs) -> None:
        """
        Add DPoP and Authorization headers to the request.
        
        Args:
            request: The prepared request to modify.
            **kwargs: Additional arguments passed to the adapter.
        """
        # Get a valid token
        token = self.token_source.token()
        
        # Parse the URL to get the path
        parsed_url = urlparse(request.url)
        target_url = parsed_url.path
        if parsed_url.query:
            target_url += f"?{parsed_url.query}"
        
        # Generate a DPoP proof for this request
        dpop_proof = self.client.generate_dpop_proof(target_url, request.method)
        
        # Add the DPoP proof to the request
        request.headers["DPoP"] = dpop_proof
        
        # Add the token to the request
        request.headers["Authorization"] = f"{token.token_type} {token.access_token}"


class DPoPHTTPClient:
    """
    An HTTP client that automatically adds DPoP proofs to requests.
    """
    
    def __init__(self, client):
        """
        Initialize the DPoP HTTP client.
        
        Args:
            client: The Client instance to use for DPoP proof generation.
        """
        self.client = client
        self.session = requests.Session()
        
        # Add the DPoP transport
        adapter = DPoPTransport(client, client.token_source)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)
    
    def request(self, method: str, url: str, **kwargs) -> requests.Response:
        """
        Make an HTTP request with DPoP proof.
        
        Args:
            method: The HTTP method to use.
            url: The URL to request.
            **kwargs: Additional arguments to pass to requests.
            
        Returns:
            The HTTP response.
        """
        return self.session.request(method, url, **kwargs)
    
    def get(self, url: str, **kwargs) -> requests.Response:
        """
        Make a GET request with DPoP proof.
        
        Args:
            url: The URL to request.
            **kwargs: Additional arguments to pass to requests.
            
        Returns:
            The HTTP response.
        """
        return self.request("GET", url, **kwargs)
    
    def post(self, url: str, **kwargs) -> requests.Response:
        """
        Make a POST request with DPoP proof.
        
        Args:
            url: The URL to request.
            **kwargs: Additional arguments to pass to requests.
            
        Returns:
            The HTTP response.
        """
        return self.request("POST", url, **kwargs)
    
    def put(self, url: str, **kwargs) -> requests.Response:
        """
        Make a PUT request with DPoP proof.
        
        Args:
            url: The URL to request.
            **kwargs: Additional arguments to pass to requests.
            
        Returns:
            The HTTP response.
        """
        return self.request("PUT", url, **kwargs)
    
    def delete(self, url: str, **kwargs) -> requests.Response:
        """
        Make a DELETE request with DPoP proof.
        
        Args:
            url: The URL to request.
            **kwargs: Additional arguments to pass to requests.
            
        Returns:
            The HTTP response.
        """
        return self.request("DELETE", url, **kwargs)
    
    def patch(self, url: str, **kwargs) -> requests.Response:
        """
        Make a PATCH request with DPoP proof.
        
        Args:
            url: The URL to request.
            **kwargs: Additional arguments to pass to requests.
            
        Returns:
            The HTTP response.
        """
        return self.request("PATCH", url, **kwargs)

File: sdks/nodejs/src/httpClient.ts
/**
 * HTTP client implementation for the Tyk FAPI SDK.
 */

import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
import { URL } from 'url';
import { TokenSource } from './tokenSource';

/**
 * An HTTP client that automatically adds DPoP proofs to requests.
 */
export class DPoPHTTPClient {
  /**
   * The Client instance to use for DPoP proof generation.
   */
  private client: any; // Avoid circular import
  
  /**
   * The TokenSource to use for token acquisition.
   */
  private tokenSource: TokenSource;
  
  /**
   * The Axios instance.
   */
  private axiosInstance: AxiosInstance;
  
  /**
   * Initialize the DPoP HTTP client.
   * 
   * @param client - The Client instance to use for DPoP proof generation.
   */
  constructor(client: any) {
    this.client = client;
    this.tokenSource = client.tokenSource;
    
    // Create an Axios instance
    this.axiosInstance = axios.create({
      timeout: 10000
    });
    
    // Add a request interceptor to add DPoP proofs
    this.axiosInstance.interceptors.request.use(
      async (config) => {
        return this.addDPoPHeaders(config);
      },
      (error) => {
        return Promise.reject(error);
      }
    );
  }
  
  /**
   * Add DPoP and Authorization headers to the request.
   * 
   * @param config - The Axios request config.
   * @returns The modified config.
   */
  private async addDPoPHeaders(config: AxiosRequestConfig): Promise<AxiosRequestConfig> {
    // Get a valid token
    const token = await this.tokenSource.getToken();
    
    // Parse the URL to get the path
    const url = new URL(config.url || '', config.baseURL || 'http://localhost');
    const targetUrl = url.pathname + url.search;
    
    // Generate a DPoP proof for this request
    const dpopProof = await this.client.generateDPoPProof(targetUrl, config.method || 'GET');
    
    // Add the DPoP proof to the request
    config.headers = config.headers || {};
    config.headers['DPoP'] = dpopProof;
    
    // Add the token to the request
    config.headers['Authorization'] = `${token.tokenType} ${token.accessToken}`;
    
    return config;
  }
  
  /**
   * Make an HTTP request with DPoP proof.
   * 
   * @param method - The HTTP method to use.
   * @param url - The URL to request.
   * @param config - Additional Axios config.
   * @returns A Promise that resolves to the HTTP response.
   */
  public async request<T = any>(
    method: string,
    url: string,
    config?: AxiosRequestConfig
  ): Promise<AxiosResponse<T>> {
    return this.axiosInstance.request<T>({
      method,
      url,
      ...config
    });
  }
  
  /**
   * Make a GET request with DPoP proof.
   * 
   * @param url - The URL to request.
   * @param config - Additional Axios config.
   * @returns A Promise that resolves to the HTTP response.
   */
  public async get<T = any>(
    url: string,
    config?: AxiosRequestConfig
  ): Promise<AxiosResponse<T>> {
    return this.request<T>('GET', url, config);
  }
  
  /**
   * Make a POST request with DPoP proof.
   * 
   * @param url - The URL to request.
   * @param data - The data to send.
   * @param config - Additional Axios config.
   * @returns A Promise that resolves to the HTTP response.
   */
  public async post<T = any>(
    url: string,
    data?: any,
    config?: AxiosRequestConfig
  ): Promise<AxiosResponse<T>> {
    return this.request<T>('POST', url, {
      ...config,
      data
    });
  }
  
  /**
   * Make a PUT request with DPoP proof.
   * 
   * @param url - The URL to request.
   * @param data - The data to send.
   * @param config - Additional Axios config.
   * @returns A Promise that resolves to the HTTP response.
   */
  public async put<T = any>(
    url: string,
    data?: any,
    config?: AxiosRequestConfig
  ): Promise<AxiosResponse<T>> {
    return this.request<T>('PUT', url, {
      ...config,
      data
    });
  }
  
  /**
   * Make a DELETE request with DPoP proof.
   * 
   * @param url - The URL to request.
   * @param config - Additional Axios config.
   * @returns A Promise that resolves to the HTTP response.
   */
  public async delete<T = any>(
    url: string,
    config?: AxiosRequestConfig
  ): Promise<AxiosResponse<T>> {
    return this.request<T>('DELETE', url, config);
  }
  
  /**
   * Make a PATCH request with DPoP proof.
   * 
   * @param url - The URL to request.
   * @param data - The data to send.
   * @param config - Additional Axios config.
   * @returns A Promise that resolves to the HTTP response.
   */
  public async patch<T = any>(
    url: string,
    data?: any,
    config?: AxiosRequestConfig
  ): Promise<AxiosResponse<T>> {
    return this.request<T>('PATCH', url, {
      ...config,
      data
    });
  }
  
  /**
   * Get the underlying Axios instance.
   * 
   * @returns The Axios instance.
   */
  public getAxiosInstance(): AxiosInstance {
    return this.axiosInstance;
  }
}

/**
 * Create a DPoP HTTP client.
 * 
 * @param client - The Client instance to use for DPoP proof generation.
 * @returns A new DPoPHTTPClient instance.
 */
export function createDPoPHTTPClient(client: any): DPoPHTTPClient {
  return new DPoPHTTPClient(client);
}